File: src/main/java/com/koushikdutta/scratch/buffers/WritableBuffers.java
package com.koushikdutta.scratch.buffers;

import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.charset.Charset;

public interface WritableBuffers {
    ByteOrder order();
    void order(ByteOrder order);

    WritableBuffers add(ReadableBuffers b);
    WritableBuffers add(ByteBuffer b);
    default WritableBuffers addAll(ByteBuffer... bb) {
        for (ByteBuffer b: bb) {
            add(b);
        }
        return this;
    }
    default WritableBuffers add(byte[] bytes) {
        return add(bytes, 0, bytes.length);
    }
    default WritableBuffers add(byte[] bytes, int offset, int length) {
        return add(ByteBuffer.wrap(bytes, offset, length));
    }

    WritableBuffers put(byte b);
    WritableBuffers putBytes(byte[] bytes);
    WritableBuffers putShort(short s);
    WritableBuffers putInt(int i);
    WritableBuffers putLong(long l);
    WritableBuffers putByteChar(char c);
    default WritableBuffers putString(String s) {
        return putString(s, null);
    }
    WritableBuffers putString(String s, Charset charset);
}

File: src/commonMain/kotlin/com.koushikdutta.scratch/buffers/WritableBuffers.kt
package com.koushikdutta.scratch.buffers

interface WritableBuffers {
    fun order(): ByteOrder
    fun order(order: ByteOrder)

    fun add(b: ReadableBuffers): WritableBuffers
    fun add(b: ByteBuffer): WritableBuffers
    fun addAll(vararg bb: ByteBuffer): WritableBuffers {
        for (b in bb) {
            add(b)
        }
        return this
    }

    fun add(bytes: ByteArray, offset: Int = 0, length: Int = bytes.size): WritableBuffers {
        return add(createByteBuffer(bytes, offset, length))
    }

    fun put(b: Byte): WritableBuffers
    fun putBytes(bytes: ByteArray): WritableBuffers
    fun putShort(s: Short): WritableBuffers
    fun putInt(i: Int): WritableBuffers
    fun putLong(l: Long): WritableBuffers
    fun putByteChar(c: Char): WritableBuffers

    fun putUtf8String(s: String): WritableBuffers
}
